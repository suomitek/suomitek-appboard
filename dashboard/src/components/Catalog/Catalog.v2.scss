.filters-menu {
  padding-left: 1.2rem;
  h5 {
    text-transform: uppercase;
    font-size: small;
    font-weight: 500;
  }
  .filter-section {
    font-weight: 500;
    margin: 0.7rem 0 0.5rem 0;
  }
}

.empty-catalog {
  text-align: center;
  width: 100%;
  height: calc(100vh - 6rem);
  // margin-top: 2rem;
  cds-icon {
    margin-top: 2rem;
    width: 4rem;
    height: 4rem;
  }
  cds-button {
    margin-top: 2rem;
  }
  p {
    margin-top: 0.6rem;
  }
}
