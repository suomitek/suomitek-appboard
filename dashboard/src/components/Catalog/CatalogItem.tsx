import * as React from "react";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import helmIcon from "../../icons/helm.svg";
import operatorIcon from "../../icons/operator-framework.svg";
import placeholder from "../../placeholder.png";
import { IRepo, IStoreState } from "../../shared/types";
import * as url from "../../shared/url";
import InfoCard from "../InfoCard";
import "./CatalogItem.css";

export interface ICatalogItem {
  id: string;
  name: string;
  version: string;
  description: string;
  cluster: string;
  namespace: string;
  icon?: string;
}

export interface IChartCatalogItem extends ICatalogItem {
  repo: IRepo;
}

export interface IOperatorCatalogItem extends ICatalogItem {
  csv: string;
}

export interface ICatalogItemProps {
  type: string;
  item: IChartCatalogItem | IOperatorCatalogItem;
}

// 3 lines description max
const MAX_DESC_LENGTH = 90;

function trimDescription(desc: string): string {
  if (desc.length > MAX_DESC_LENGTH) {
    // Trim to the last word under the max length
    return desc.substr(0, desc.lastIndexOf(" ", MAX_DESC_LENGTH)).concat("...");
  }
  return desc;
}

const CatalogItem: React.SFC<ICatalogItemProps> = props => {
  if (props.type === "operator") {
    const item = props.item as IOperatorCatalogItem;
    return OperatorCatalogItem(item);
  } else {
    const item = props.item as IChartCatalogItem;
    return ChartCatalogItem(item);
  }
};

const OperatorCatalogItem: React.SFC<IOperatorCatalogItem> = props => {
  const { icon, name, csv, version, description, cluster, namespace, id } = props;
  const iconSrc = icon || placeholder;
  // Cosmetic change, remove the version from the csv name
  const csvName = props.csv.split(".v")[0];
  const tag1 = <span>{csvName}</span>;
  const link = url.app.operatorInstances.new(cluster, namespace, csv, id);
  const subIcon = operatorIcon;
  const descriptionC = (
    <div className="ListItem__content__description">{trimDescription(description)}</div>
  );
  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || "-"}
      icon={iconSrc}
      description={descriptionC}
      tag1Content={tag1}
      tag1Class={""}
      subIcon={subIcon}
    />
  );
};

const ChartCatalogItem: React.SFC<IChartCatalogItem> = props => {
  const { icon, name, repo, version, description, namespace, id } = props;
  const iconSrc = icon || placeholder;
  const cluster = useSelector((state: IStoreState) => state.clusters.currentCluster);
  const tag1 = (
    <Link
      className="ListItem__content__info_tag_link"
      to={url.app.repo(cluster, namespace, repo.name)}
    >
      {repo.name}
    </Link>
  );
  const link = url.app.charts.get(cluster, namespace, name, repo || ({} as IRepo));
  const subIcon = helmIcon;

  const descriptionC = (
    <div className="ListItem__content__description">{trimDescription(description)}</div>
  );
  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || "-"}
      icon={iconSrc}
      description={descriptionC}
      tag1Content={tag1}
      tag1Class={repo ? repo.name : ""}
      subIcon={subIcon}
    />
  );
};

export default CatalogItem;
