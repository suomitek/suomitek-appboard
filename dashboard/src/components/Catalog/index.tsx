import * as React from "react";

import { ICatalogProps } from "./Catalog";

const Catalog = React.lazy(() => import("./Catalog"));
const CatalogV2 = React.lazy(() => import("./Catalog.v2"));

interface ICatalogSelectorProps extends ICatalogProps {
  UI: string;
}

const CatalogSelector: React.FC<ICatalogSelectorProps> = props => (
  <React.Suspense fallback={null}>
    {props.UI === "clarity" ? <CatalogV2 {...props} /> : <Catalog {...props} />}
  </React.Suspense>
);

export default CatalogSelector;
