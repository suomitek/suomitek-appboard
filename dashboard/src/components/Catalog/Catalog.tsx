import { RouterAction } from "connected-react-router";
import * as React from "react";
import { Link } from "react-router-dom";

import { IFeatureFlags } from "shared/Config";
import { app } from "shared/url";
import { ForbiddenError, IChart, IChartState, IClusterServiceVersion } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import { CardGrid } from "../Card";
import { MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import CatalogItem, {
  ICatalogItemProps,
  IChartCatalogItem,
  IOperatorCatalogItem,
} from "./CatalogItem";

export interface ICatalogProps {
  charts: IChartState;
  repo: string;
  filter: string;
  fetchCharts: (namespace: string, repo: string) => void;
  pushSearchFilter: (filter: string) => RouterAction;
  cluster: string;
  namespace: string;
  kubeappsNamespace: string;
  getCSVs: (namespace: string) => void;
  csvs: IClusterServiceVersion[];
  featureFlags: IFeatureFlags;
}

interface ICatalogState {
  filter: string;
  listCharts: boolean;
  listOperators: boolean;
}

class Catalog extends React.Component<ICatalogProps, ICatalogState> {
  public state: ICatalogState = {
    filter: "",
    listCharts: true,
    listOperators: true,
  };

  public componentDidMount() {
    const { repo, fetchCharts, filter, namespace, getCSVs, featureFlags } = this.props;
    this.setState({ filter });
    fetchCharts(namespace, repo);
    if (featureFlags.operators) {
      getCSVs(namespace);
    }
  }

  public componentDidUpdate(prevProps: ICatalogProps) {
    if (this.props.filter !== prevProps.filter) {
      this.setState({ filter: this.props.filter });
    }
    if (this.props.repo !== prevProps.repo || this.props.namespace !== prevProps.namespace) {
      this.props.fetchCharts(this.props.namespace, this.props.repo);
    }
    if (this.props.namespace !== prevProps.namespace && this.props.featureFlags.operators) {
      this.props.getCSVs(this.props.namespace);
    }
  }

  public render() {
    const {
      charts: {
        isFetching,
        selected: { error },
        items: allItems,
      },
      cluster,
      namespace,
      pushSearchFilter,
      csvs,
    } = this.props;
    const { listCharts, listOperators } = this.state;
    if (error) {
      const isForbidden = error.constructor === ForbiddenError;
      return (
        <MessageAlert
          level={"error"}
          children={
            <div>
              <h5>Unable to fetch catalog</h5>
              There was an error fetching the catalog.
              {isForbidden && " Please choose a namespace above to which you have access."}
            </div>
          }
        />
      );
    }
    if (!isFetching && allItems.length === 0 && csvs.length === 0) {
      return (
        <MessageAlert
          level={"warning"}
          children={
            <div>
              <h5>Charts not found.</h5>
              Manage your Helm chart repositories in Suomitek-appboard by visiting the{" "}
              <Link to={app.config.apprepositories(cluster, namespace)}>
                App repositories configuration
              </Link>{" "}
              page.
            </div>
          }
        />
      );
    }
    const filteredCharts = this.filteredCharts(allItems);
    const filteredCSVs = this.shouldRenderOperators() ? this.filteredCSVs(csvs) : [];
    const catalogItems = this.getCatalogItems(filteredCharts, filteredCSVs);
    const items = catalogItems.map(c => {
      const keyComponent =
        c.type === "operator"
          ? (c.item as IOperatorCatalogItem).csv
          : (c.item as IChartCatalogItem).repo.name;
      return (
        <CatalogItem type={c.type} key={`${c.type}/${keyComponent}/${c.item.name}`} item={c.item} />
      );
    });
    return (
      <section className="Catalog">
        <PageHeader>
          <h1>Catalog</h1>
          <SearchFilter
            className="margin-l-big"
            placeholder="search charts..."
            onChange={this.handleFilterQueryChange}
            value={this.state.filter}
            onSubmit={pushSearchFilter}
          />
        </PageHeader>
        <LoadingWrapper loaded={!isFetching}>
          <div className="row">
            {this.shouldRenderOperators() && (
              <div className="col-2">
                <div className="margin-b-normal">
                  <span>
                    <b>Type:</b>
                  </span>
                </div>
                <div>
                  <label className="checkbox" key="listcharts">
                    <input type="checkbox" checked={listCharts} onChange={this.toggleListCharts} />
                    <span>Charts</span>
                  </label>
                </div>
                <div>
                  <label className="checkbox" key="listoperators">
                    <input
                      type="checkbox"
                      checked={listOperators}
                      onChange={this.toggleListOperators}
                    />
                    <span>Operators</span>
                  </label>
                </div>
              </div>
            )}
            <div className={this.shouldRenderOperators() ? "col-10" : "col-12"}>
              <CardGrid>{items}</CardGrid>
            </div>
          </div>
        </LoadingWrapper>
      </section>
    );
  }

  private filteredCharts = (charts: IChart[]) => {
    const { filter, listCharts } = this.state;
    if (!listCharts) {
      return [];
    }
    return charts.filter(c => new RegExp(escapeRegExp(filter), "i").test(c.id));
  };

  private filteredCSVs(csvs: IClusterServiceVersion[]) {
    const { filter, listOperators } = this.state;
    if (!listOperators) {
      return [];
    }
    return csvs.filter(c => new RegExp(escapeRegExp(filter), "i").test(c.metadata.name));
  }

  private getCatalogItems(charts: IChart[], csvs: IClusterServiceVersion[]): ICatalogItemProps[] {
    let result: ICatalogItemProps[] = [];
    charts.forEach(c => {
      result = result.concat({
        type: "chart",
        item: {
          id: c.id,
          name: c.attributes.name,
          icon: c.attributes.icon ? `api/assetsvc/${c.attributes.icon}` : undefined,
          version: c.relationships.latestChartVersion.data.app_version,
          description: c.attributes.description,
          repo: c.attributes.repo,
          namespace: this.props.namespace,
          cluster: this.props.cluster,
        },
      });
    });
    csvs.forEach(csv => {
      csv.spec.customresourcedefinitions.owned.forEach(crd => {
        result = result.concat({
          type: "operator",
          item: {
            id: crd.name,
            name: crd.displayName,
            icon: `data:${csv.spec.icon[0].mediatype};base64,${csv.spec.icon[0].base64data}`,
            version: crd.version,
            description: crd.description,
            csv: csv.metadata.name,
            namespace: this.props.namespace,
            cluster: this.props.cluster,
          },
        });
      });
    });
    return result.sort((a, b) => (a.item.name.toLowerCase() > b.item.name.toLowerCase() ? 1 : -1));
  }

  private toggleListCharts = () => {
    this.setState({ listCharts: !this.state.listCharts });
  };

  private toggleListOperators = () => {
    this.setState({ listOperators: !this.state.listOperators });
  };

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };

  // Render Operators only if there are Operators to render and if we are not
  // filtering by repo
  private shouldRenderOperators = () => {
    const { csvs, repo } = this.props;
    return csvs.length > 0 && !repo;
  };
}

export default Catalog;
