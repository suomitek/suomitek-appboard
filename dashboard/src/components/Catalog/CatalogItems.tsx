import React, { useMemo } from "react";
import { getIcon } from "shared/Operators";
import { IChart, IClusterServiceVersion } from "shared/types";
import CatalogItem, { ICatalogItemProps } from "./CatalogItem.v2";

interface ICatalogItemsProps {
  charts: IChart[];
  csvs: IClusterServiceVersion[];
  cluster: string;
  namespace: string;
}

export default function CatalogItems({ charts, csvs, cluster, namespace }: ICatalogItemsProps) {
  const chartItems: ICatalogItemProps[] = useMemo(
    () =>
      charts.map(c => {
        return {
          type: "chart",
          id: `chart/${c.attributes.repo.name}/${c.id}`,
          item: {
            id: c.id,
            name: c.attributes.name,
            icon: c.attributes.icon ? `api/assetsvc/${c.attributes.icon}` : undefined,
            version: c.relationships.latestChartVersion.data.app_version,
            description: c.attributes.description,
            repo: c.attributes.repo,
            cluster,
            namespace,
          },
        };
      }),
    [charts, cluster, namespace],
  );
  const crdItems: ICatalogItemProps[] = useMemo(
    () =>
      csvs
        .map(csv => {
          if (csv.spec.customresourcedefinitions?.owned) {
            return csv.spec.customresourcedefinitions.owned.map(crd => {
              return {
                type: "operator",
                id: `operator/${csv.metadata.name}/${crd.name}`,
                item: {
                  id: crd.name,
                  name: crd.displayName || crd.name,
                  icon: getIcon(csv),
                  version: crd.version,
                  description: crd.description,
                  csv: csv.metadata.name,
                  cluster,
                  namespace,
                },
              };
            });
          } else {
            return [];
          }
        })
        .flat(),
    [csvs, cluster, namespace],
  );

  const sortedItems = chartItems
    .concat(crdItems)
    .sort((a, b) => (a.item.name.toLowerCase() > b.item.name.toLowerCase() ? 1 : -1));

  if (sortedItems.length === 0) {
    return <p>No application matches the current filter.</p>;
  }
  return (
    <>
      {sortedItems.map(i => (
        <CatalogItem key={i.id} {...i} />
      ))}
    </>
  );
}
