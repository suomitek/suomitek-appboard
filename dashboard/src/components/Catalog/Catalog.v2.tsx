import { RouterAction } from "connected-react-router";
import { flatten, get, intersection, uniq } from "lodash";
import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { CdsIcon } from "../Clarity/clarity";

import FilterGroup from "components/FilterGroup/FilterGroup";
import Alert from "components/js/Alert";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { IFeatureFlags } from "shared/Config";
import { IChartState, IClusterServiceVersion } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import CardGrid from "../Card/CardGrid.v2";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper.v2";
import PageHeader from "../PageHeader/PageHeader.v2";
import SearchFilter from "../SearchFilter/SearchFilter.v2";

import { CdsButton } from "components/Clarity/clarity";
import { app } from "shared/url";
import "./Catalog.v2.css";
import CatalogItems from "./CatalogItems";

function categoryToReadable(category: string) {
  return category === "" ? "Unknown" : category.replace(/([a-z])([A-Z][a-z])/g, "$1 $2").trimLeft();
}

function getOperatorCategories(c: IClusterServiceVersion): string[] {
  return get(c, "metadata.annotations.categories", "")
    .split(",")
    .map((category: string) => categoryToReadable(category));
}

interface ICatalogProps {
  charts: IChartState;
  repo: string;
  filter: string;
  fetchCharts: (namespace: string, repo: string) => void;
  pushSearchFilter: (filter: string) => RouterAction;
  cluster: string;
  namespace: string;
  kubeappsNamespace: string;
  getCSVs: (namespace: string) => void;
  csvs: IClusterServiceVersion[];
  featureFlags: IFeatureFlags;
}

function Catalog(props: ICatalogProps) {
  const {
    charts: {
      isFetching,
      selected: { error },
      items: charts,
    },
    fetchCharts,
    cluster,
    namespace,
    pushSearchFilter,
    getCSVs,
    csvs,
    repo,
    filter: propsFilter,
  } = props;
  const [searchFilter, setSearchFilter] = useState(propsFilter);
  const [typeFilter, setTypeFilter] = useState([] as string[]);
  const [repoFilter, setRepoFilter] = useState([] as string[]);
  const [categoryFilter, setCategoryFilter] = useState([] as string[]);
  const [operatorProviderFilter, setOperatorProviderFilter] = useState([] as string[]);

  const allRepos = uniq(charts.map(c => c.attributes.repo.name));
  const allProviders = uniq(csvs.map(c => c.spec.provider.name));
  const allCategories = uniq(
    charts
      .map(c => categoryToReadable(c.attributes.category))
      .concat(flatten(csvs.map(c => getOperatorCategories(c)))),
  ).sort();

  useEffect(() => {
    fetchCharts(namespace, repo);
    getCSVs(namespace);
  }, [namespace, repo, fetchCharts, getCSVs]);

  useEffect(() => {
    setSearchFilter(propsFilter);
  }, [propsFilter]);

  const filteredCharts = charts
    .filter(() => typeFilter.length === 0 || typeFilter.includes("Charts"))
    .filter(() => operatorProviderFilter.length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.id))
    .filter(c => repoFilter.length === 0 || repoFilter.includes(c.attributes.repo.name))
    .filter(
      c =>
        categoryFilter.length === 0 ||
        categoryFilter.includes(categoryToReadable(c.attributes.category)),
    );
  const filteredCSVs = csvs
    .filter(() => typeFilter.length === 0 || typeFilter.includes("Operators"))
    .filter(() => repoFilter.length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.metadata.name))
    .filter(
      c =>
        operatorProviderFilter.length === 0 ||
        operatorProviderFilter.includes(c.spec.provider.name),
    )
    .filter(
      c =>
        categoryFilter.length === 0 ||
        intersection(categoryFilter, getOperatorCategories(c)).length,
    );

  return (
    <section>
      <PageHeader>
        <Row>
          <h1>Catalog</h1>
          <SearchFilter
            key="searchFilter"
            placeholder="search charts..."
            onChange={setSearchFilter}
            value={searchFilter}
            onSubmit={pushSearchFilter}
          />
        </Row>
      </PageHeader>
      <LoadingWrapper loaded={!isFetching}>
        {error && (
          <Alert theme="danger">Found en error fetching the catalog: {error.message}</Alert>
        )}
        {charts.length === 0 && csvs.length === 0 ? (
          <div className="empty-catalog">
            <CdsIcon shape="bundle" />
            <p>The current catalog is empty.</p>
            <p>
              Manage your Helm chart repositories in Suomitek-appboard by visiting the App repositories
              configuration page.
            </p>
            <Link to={app.config.apprepositories(cluster, namespace)}>
              <CdsButton>Manage App Repositories</CdsButton>
            </Link>
          </div>
        ) : (
          <Row>
            <Column span={2}>
              <div className="filters-menu">
                <h5>Filters</h5>
                {csvs.length > 0 && (
                  <div className="filter-section">
                    <label>Application Type:</label>
                    <FilterGroup
                      name="apptype"
                      options={["Operators", "Charts"]}
                      onChange={setTypeFilter}
                    />
                  </div>
                )}
                {allCategories.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Category:</label>
                    <FilterGroup
                      name="category"
                      options={allCategories}
                      onChange={setCategoryFilter}
                    />
                  </div>
                )}
                {allRepos.length > 0 && (
                  <div className="filter-section">
                    <label>Application Repository:</label>
                    <FilterGroup name="apprepo" options={allRepos} onChange={setRepoFilter} />
                  </div>
                )}
                {allProviders.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Operator Provider:</label>
                    <FilterGroup
                      name="operator-provider"
                      options={allProviders}
                      onChange={setOperatorProviderFilter}
                    />
                  </div>
                )}
              </div>
            </Column>
            <Column span={10}>
              <CardGrid>
                <CatalogItems
                  charts={filteredCharts}
                  csvs={filteredCSVs}
                  cluster={cluster}
                  namespace={namespace}
                />
              </CardGrid>
            </Column>
          </Row>
        )}
      </LoadingWrapper>
    </section>
  );
}

export default Catalog;
