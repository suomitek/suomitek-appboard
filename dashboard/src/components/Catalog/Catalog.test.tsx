import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../shared/specs";
import { ForbiddenError, IChart, IChartState } from "../../shared/types";
import { CardGrid } from "../Card";
import { MessageAlert } from "../ErrorAlert";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import Catalog from "./Catalog";
import CatalogItem from "./CatalogItem";

const defaultChartState = {
  isFetching: false,
  selected: {} as IChartState["selected"],
  deployed: {} as IChartState["deployed"],
  items: [],
  updatesInfo: {},
} as IChartState;
const defaultProps = {
  charts: defaultChartState,
  repo: "",
  filter: "",
  fetchCharts: jest.fn(),
  pushSearchFilter: jest.fn(),
  cluster: "default",
  namespace: "kubeapps",
  kubeappsNamespace: "kubeapps",
  csvs: [],
  getCSVs: jest.fn(),
  featureFlags: { operators: false, ui: "hex" },
};

it("propagates the filter from the props", () => {
  const wrapper = shallow(<Catalog {...defaultProps} filter="foo" />);
  expect(wrapper.state("filter")).toBe("foo");
});

it("reloads charts when the repo changes", () => {
  const fetchCharts = jest.fn();
  const wrapper = shallow(<Catalog {...defaultProps} fetchCharts={fetchCharts} />);
  wrapper.setProps({ ...defaultProps, fetchCharts, repo: "bitnami" });
  expect(fetchCharts.mock.calls.length).toBe(2);
  expect(fetchCharts.mock.calls[1]).toEqual(["kubeapps", "bitnami"]);
});

it("updates the filter from props", () => {
  const wrapper = shallow(<Catalog {...defaultProps} />);
  wrapper.setProps({ filter: "foo" });
  expect(wrapper.state("filter")).toBe("foo");
});

it("keeps the filter from the state", () => {
  const wrapper = shallow(<Catalog {...defaultProps} />);
  expect(wrapper.state("filter")).toBe("");
  wrapper.setState({ filter: "foo" });
  expect(wrapper.state("filter")).toBe("foo");
});

describe("componentDidMount", () => {
  it("retrieves csvs in the namespace", () => {
    const getCSVs = jest.fn();
    const namespace = "foo";
    shallow(
      <Catalog
        {...defaultProps}
        getCSVs={getCSVs}
        namespace={namespace}
        featureFlags={{ ...defaultProps.featureFlags, operators: true }}
      />,
    );
    expect(getCSVs).toHaveBeenCalledWith(namespace);
  });
});

describe("componentDidUpdate", () => {
  it("re-fetches csvs if the namespace changes", () => {
    const getCSVs = jest.fn();
    const wrapper = shallow(
      <Catalog
        {...defaultProps}
        getCSVs={getCSVs}
        featureFlags={{ ...defaultProps.featureFlags, operators: true }}
      />,
    );
    wrapper.setProps({ namespace: "a-different-one" });
    expect(getCSVs).toHaveBeenCalledWith("a-different-one");
  });
});

describe("renderization", () => {
  context("when no charts", () => {
    it("should render an error", () => {
      const wrapper = shallow(<Catalog {...defaultProps} />);
      expect(wrapper.find(MessageAlert)).toExist();
      expect(wrapper.find(".Catalog")).not.toExist();
      expect(
        wrapper
          .find(MessageAlert)
          .children()
          .text(),
      ).toContain("Manage your Helm chart repositories");
      expect(wrapper).toMatchSnapshot();
    });
  });

  context("when there is an error fetching charts", () => {
    it("should render a generic error", () => {
      const props = {
        ...defaultProps,
        charts: {
          ...defaultProps.charts,
          selected: {
            ...defaultProps.charts.selected,
            error: new Error("Bang!"),
          },
        },
      };
      const wrapper = shallow(<Catalog {...props} />);
      expect(wrapper.find(MessageAlert)).toExist();
      expect(wrapper.find(".Catalog")).not.toExist();
      const errText = wrapper
        .find(MessageAlert)
        .children()
        .text();
      expect(errText).toContain("Unable to fetch catalog");
      expect(errText).not.toContain("Please choose a namespace");
      expect(wrapper).toMatchSnapshot();
    });

    it("should render a specific error for a forbidden error", () => {
      const props = {
        ...defaultProps,
        charts: {
          ...defaultProps.charts,
          selected: {
            ...defaultProps.charts.selected,
            error: new ForbiddenError("Bang!"),
          },
        },
      };
      const wrapper = shallow(<Catalog {...props} />);
      expect(wrapper.find(MessageAlert)).toExist();
      expect(wrapper.find(".Catalog")).not.toExist();
      const errText = wrapper
        .find(MessageAlert)
        .children()
        .text();
      expect(errText).toContain("Unable to fetch catalog");
      expect(errText).toContain("Please choose a namespace");
      expect(wrapper).toMatchSnapshot();
    });
  });

  context("when fetching apps", () => {
    itBehavesLike("aLoadingComponent", {
      component: Catalog,
      props: { ...defaultProps, charts: { isFetching: true, items: [], selected: {} } },
    });
  });

  context("when charts available", () => {
    const chartState = {
      isFetching: false,
      selected: {} as IChartState["selected"],
      items: [
        {
          id: "foo",
          attributes: {
            name: "foo",
            description: "",
            repo: { name: "foo", namespace: "chart-namespace" },
          },
          relationships: { latestChartVersion: { data: { app_version: "v1.0.0" } } },
        } as IChart,
        {
          id: "bar",
          attributes: {
            name: "bar",
            description: "",
            repo: { name: "bar", namespace: "chart-namespace" },
          },
          relationships: { latestChartVersion: { data: { app_version: "v2.0.0" } } },
        } as IChart,
      ],
    } as IChartState;

    it("should render the list of charts", () => {
      const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} />);

      expect(wrapper.find(MessageAlert)).not.toExist();
      expect(wrapper.find(PageHeader)).toExist();
      expect(wrapper.find(SearchFilter)).toExist();

      const cardGrid = wrapper.find(CardGrid);
      expect(cardGrid).toExist();
      expect(cardGrid.children().length).toBe(chartState.items.length);
      const expectedItem1 = {
        description: "",
        id: "bar",
        name: "bar",
        namespace: "kubeapps",
        cluster: "default",
        repo: {
          name: "bar",
          namespace: "chart-namespace",
        },
        version: "v2.0.0",
      };
      const expectedItem2 = {
        description: "",
        id: "foo",
        name: "foo",
        namespace: "kubeapps",
        cluster: "default",
        repo: {
          name: "foo",
          namespace: "chart-namespace",
        },
        version: "v1.0.0",
      };
      expect(
        cardGrid
          .children()
          .at(0)
          .props().item,
      ).toEqual(expectedItem1);
      expect(
        cardGrid
          .children()
          .at(1)
          .props().item,
      ).toEqual(expectedItem2);
      expect(wrapper).toMatchSnapshot();
      // If there are no csvs, there shouldn't be columns
      expect(wrapper.find(".col-10")).not.toExist();
    });

    it("should filter apps", () => {
      // Filter "foo" app
      const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} filter="foo" />);

      const cardGrid = wrapper.find(CardGrid);
      expect(cardGrid).toExist();
      expect(cardGrid.children().length).toBe(1);
      const expectedItem = {
        description: "",
        id: "foo",
        name: "foo",
        namespace: "kubeapps",
        cluster: "default",
        repo: {
          name: "foo",
          namespace: "chart-namespace",
        },
        version: "v1.0.0",
      };
      expect(
        cardGrid
          .children()
          .at(0)
          .props().item,
      ).toEqual(expectedItem);
    });

    describe("when operators available", () => {
      const csvs = [
        {
          metadata: {
            name: "test-csv",
          },
          spec: {
            icon: [{ base64data: "data", mediatype: "img/png" }],
            customresourcedefinitions: {
              owned: [
                {
                  name: "foo-cluster",
                  displayName: "foo-cluster",
                  version: "v1.0.0",
                  description: "a meaningful description",
                },
              ],
            },
          },
        } as any,
      ];

      it("shows the list of charts and operators", () => {
        const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} csvs={csvs} />);
        const cardGrid = wrapper.find(CardGrid);
        expect(cardGrid).toExist();
        expect(cardGrid.children().length).toBe(chartState.items.length + csvs.length);

        const expectedItem = {
          csv: "test-csv",
          description: "a meaningful description",
          icon: "data:img/png;base64,data",
          id: "foo-cluster",
          name: "foo-cluster",
          namespace: "kubeapps",
          version: "v1.0.0",
        };
        const csvCard = cardGrid
          .find(CatalogItem)
          .findWhere(c => c.prop("item").id === "foo-cluster");
        expect(csvCard).toExist();
        expect(csvCard.prop("item")).toMatchObject(expectedItem);
        // If there are no csvs, there should be a column for the cardgrid
        expect(wrapper.find(".col-10")).toExist();
        // The list should be ordered
        expect(cardGrid.children().map(c => c.props().item.id)).toEqual([
          "bar",
          "foo",
          "foo-cluster",
        ]);
      });

      it("should filter out charts or operators when requested", () => {
        const wrapper = shallow(<Catalog {...defaultProps} charts={chartState} csvs={csvs} />);

        wrapper.setState({ listCharts: false });
        let cardGrid = wrapper.find(CardGrid);
        expect(cardGrid).toExist();
        expect(cardGrid.children().length).toBe(csvs.length);

        wrapper.setState({ listOperators: false, listCharts: true });
        cardGrid = wrapper.find(CardGrid);
        expect(cardGrid).toExist();
        expect(cardGrid.children().length).toBe(chartState.items.length);
      });

      it("should show operators if there are no charts available", () => {
        const wrapper = shallow(<Catalog {...defaultProps} csvs={csvs} />);
        expect(wrapper.find(MessageAlert)).not.toExist();
        expect(wrapper.find(".Catalog")).toExist();
      });

      it("should skip operators if the repo filter is used", () => {
        const wrapper = shallow(
          <Catalog {...defaultProps} charts={chartState} csvs={csvs} repo="test" />,
        );
        const cardGrid = wrapper.find(CardGrid);
        expect(cardGrid).toExist();
        expect(cardGrid.children().length).toBe(chartState.items.length);
      });
    });
  });
});
