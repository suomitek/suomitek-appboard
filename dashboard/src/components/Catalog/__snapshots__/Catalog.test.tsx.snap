// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderization when charts available should render the list of charts 1`] = `
<section
  className="Catalog"
>
  <PageHeader>
    <h1>
      Catalog
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[MockFunction]}
      placeholder="search charts..."
      value=""
    />
  </PageHeader>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <CardGrid>
          <CatalogItem
            item={
              Object {
                "cluster": "default",
                "description": "",
                "icon": undefined,
                "id": "bar",
                "name": "bar",
                "namespace": "kubeapps",
                "repo": Object {
                  "name": "bar",
                  "namespace": "chart-namespace",
                },
                "version": "v2.0.0",
              }
            }
            key="chart/bar/bar"
            type="chart"
          />
          <CatalogItem
            item={
              Object {
                "cluster": "default",
                "description": "",
                "icon": undefined,
                "id": "foo",
                "name": "foo",
                "namespace": "kubeapps",
                "repo": Object {
                  "name": "foo",
                  "namespace": "chart-namespace",
                },
                "version": "v1.0.0",
              }
            }
            key="chart/foo/foo"
            type="chart"
          />
        </CardGrid>
      </div>
    </div>
  </LoadingWrapper>
</section>
`;

exports[`renderization when fetching apps loading spinner matches the snapshot 1`] = `
<section
  className="Catalog"
>
  <PageHeader>
    <h1>
      Catalog
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[MockFunction]}
      placeholder="search charts..."
      value=""
    />
  </PageHeader>
  <LoadingWrapper
    loaded={false}
    type={0}
  >
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <CardGrid />
      </div>
    </div>
  </LoadingWrapper>
</section>
`;

exports[`renderization when no charts should render an error 1`] = `
<MessageAlertPage
  level="warning"
>
  <div>
    <h5>
      Charts not found.
    </h5>
    Manage your Helm chart repositories in Kubeapps by visiting the
     
    <Link
      to="/c/default/ns/kubeapps/config/repos"
    >
      App repositories configuration
    </Link>
     
    page.
  </div>
</MessageAlertPage>
`;

exports[`renderization when there is an error fetching charts should render a generic error 1`] = `
<MessageAlertPage
  level="error"
>
  <div>
    <h5>
      Unable to fetch catalog
    </h5>
    There was an error fetching the catalog.
  </div>
</MessageAlertPage>
`;

exports[`renderization when there is an error fetching charts should render a specific error for a forbidden error 1`] = `
<MessageAlertPage
  level="error"
>
  <div>
    <h5>
      Unable to fetch catalog
    </h5>
    There was an error fetching the catalog.
     Please choose a namespace above to which you have access.
  </div>
</MessageAlertPage>
`;
